Spigot-Server-Patches/0180-InventoryMoveItemEvent-improvements.patch

```diff
@@ -0,0 +1,165 @@
++                        // this.getWorld().getServer().getPluginManager().callEvent(event);
++                        // Paper end
+                         if (event.isCancelled()) {
+                             this.setItem(i, itemstack);
+                             this.setCooldown(world.spigotConfig.hopperTransfer); // Spigot
+@@ -420,8 +431,18 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
+             }
+ 
+             InventoryMoveItemEvent event = new InventoryMoveItemEvent(sourceInventory, oitemstack.clone(), ihopper.getOwner().getInventory(), false);
+-
+-            ihopper.getWorld().getServer().getPluginManager().callEvent(event);
++            // Paper start - InventoryMoveItemEvent improvements
++            if (ihopper.getWorld().paperConfig.callMoveItemEvent) {
++                boolean acceptsItem = willAcceptItem(iinventory, CraftItemStack.asNMSCopy(oitemstack), enumdirection);
++                if (ihopper.getWorld().paperConfig.callMoveItemEventWithoutMove || acceptsItem) {
++                    if (ihopper.getWorld().paperConfig.callMoveEventCancelled && !acceptsItem) {
++                        event.setCancelled(true);
++                    }
++                    ihopper.getWorld().getServer().getPluginManager().callEvent(event);
++                }
++            }
++            //ihopper.getWorld().getServer().getPluginManager().callEvent(event);
++            // Paper end
+             if (event.isCancelled()) {
+                 iinventory.setItem(i, itemstack1);
+ 
+@@ -452,6 +473,46 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
+         return false;
+     }
+ 
++    // Paper start - method to check whether or not an inventory will accept the provided item. This should match Vanilla behaviour
++    // Vanilla method: addItem(IInventory, ItemsStack, EnumDirection)
++    static boolean willAcceptItem(IInventory iinventory, ItemStack itemstack, EnumDirection enumdirection) {
++        if (iinventory instanceof IWorldInventory && enumdirection != null) {
++            IWorldInventory iworldinventory = (IWorldInventory) iinventory;
++            int[] aint = iworldinventory.getSlotsForFace(enumdirection);
++
++            for (int i = 0; i < aint.length; ++i) {
++                if (enoughRoomInSlot(iinventory, itemstack, aint[i], enumdirection)) {
++                    return true;
++                }
++            }
++        } else {
++            int j = iinventory.getSize();
++
++            for (int k = 0; k < j; ++k) {
++                if (enoughRoomInSlot(iinventory, itemstack, k, enumdirection)) {
++                    return true;
++                }
++            }
++        }
++        return false;
++    }
++
++    // Vanilla method: c(IInventory, ItemStack, int, EnumDirection)
++    private static boolean enoughRoomInSlot(IInventory iinventory, ItemStack itemstack, int i, EnumDirection enumdirection) {
```

Comment:
There is precedent to do so, I'd rather see it split than a small portion copied. Unless that split has severe implications on readability. This will also help us maintain it.